http: Fix HTTP server shutdown

promag opened the pull request:
Fixes #11777. Reverts #11006. Replaces #13501.

With this change the HTTP server will exit gracefully, meaning that all requests will finish processing and sending the response, even if this means to wait more than 2 seconds (current time allowed to exit the event loop).

Another small change is that connections are accepted even when the server is stopping, but HTTP requests are rejected. This can be improved later, especially if chunked replies are implemented.

Briefly, before this PR, this is the order or events when a request arrives (RPC `stop`):
 1. `bufferevent_disable(..., EV_READ)`
 2. `StartShutdown()`
 3. `evhttp_del_accept_socket(...)`
 4. `ThreadHTTP` terminates (event loop exits) because there are no active or pending events thanks to 1. and 3.
 5. client doesn't get the response thanks to 4.

This can be verified by applying
```diff
     // Event loop will exit after current HTTP requests have been handled, so
     // this reply will get back to the client.
     StartShutdown();
+    MilliSleep(2000);
     return "Bitcoin server stopping";
 }
```
and checking the log output:
```
    Received a POST request for / from 127.0.0.1:62443
    ThreadRPCServer method=stop user=__cookie__
    Interrupting HTTP server
**  Exited http event loop
    Interrupting HTTP RPC server
    Interrupting RPC
    tor: Thread interrupt
    Shutdown: In progress...
    torcontrol thread exit
    Stopping HTTP RPC server
    addcon thread exit
    opencon thread exit
    Unregistering HTTP handler for / (exactmatch 1)
    Unregistering HTTP handler for /wallet/ (exactmatch 0)
    Stopping RPC
    RPC stopped.
    Stopping HTTP server
    Waiting for HTTP worker threads to exit
    msghand thread exit
    net thread exit

    ... sleep 2 seconds ...

    Waiting for HTTP event thread to exit
    Stopped HTTP server
```

For this reason point 3. is moved right after all HTTP workers quit. In that moment HTTP replies are queued in the event loop which keeps spinning util all connections are closed. In order to trigger the server side close with keep alive connections (implicit in HTTP/1.1) the header `Connection: close` is sent if shutdown was requested. This can be tested by
```
bitcoind -regtest
nc localhost 18443
POST / HTTP/1.1
Authorization: Basic ...
Content-Type: application/json
Content-Length: 44

{"jsonrpc": "2.0","method":"stop","id":123}
```

Summing up, this PR:
 - removes explicit event loop exit — event loop exits once there are no active or pending events
 - changes the moment the listening sockets are removed — explained above
 - sends header `Connection: close` on active requests when shutdown was requested which is relevant when it's a persistent connection (default in HTTP 1.1) — libevent is aware of this header and closes the connection gracefully
 - removes event loop explicit break after 2 seconds timeout

jnewbery commented:
@promag can you explain how this is different from #13501 and what the reasoning was for closing that PR?

MarcoFalke commented:
This is now the trillionth attempt in fixing the issue. Could you provide some rationale how this is different from the other approaches and why it should be preferred over the other ones?

ryanofsky commented:
This is simpler than #13501, and if it works I think it is preferable. Instead of trying to wait for all reply data to be sent before stopping the event loop and calling `evhttp_free`, it lets `evhttp_free` worry about this and stop the event loop itself. Assuming `http_free` does the right thing, this is a cleaner and less invasive.

I am curious what exactly the behavior of `evhttp_free` is. For example if we are in the middle of sending a huge reply to a slow client, and the server is shut down, how long does this wait (if it waits at all) to send all the data before shutting down? IIUC #13501 would wait as long as it took, and currently we wait for 2 seconds.

ryanofsky commented:
It seems like `evhttp_free` just immediately calls `shutdown` and `close` on all open connections, so I would expect this change not to always fix #11777. 

https://github.com/libevent/libevent/blob/9afe7a6c12c112a6cbc49cf1bd684781275b2579/http.c#L1243
https://github.com/libevent/libevent/blob/9afe7a6c12c112a6cbc49cf1bd684781275b2579/http.c#L3721

But maybe the `WriteHeader("Connection", "close")` call added to `HTTPRequest::WriteReply` improves things in some way, so even if there is a race condition, it won't trigger in practice?

promag commented:
@ryanofsky right, can't call `evhttp_free` with ongoing requests.

Updated code and PR description, should be more clear of the current problem and the approach used.

promag committed:
http: Send "Connection: close" header if shutdown is requested

Sending the header "Connection: close" makes libevent close persistent
connections (implicit with HTTP 1.1) which cleans the event base when
shutdown is requested.

promag committed:
http: Unlisten sockets after all workers quit

This (almost) move only ensures the event base loop doesn't exit before
HTTP worker threads exit. This way events registered by HTTP workers are
processed and not discarded.

promag committed:
http: Remove unnecessary event_base_loopexit call

Let event base loop exit cleanly by processing all active and pending
events. The call is no longer necessary because closing persistent
connections is now properly handled.

promag committed:
http: Remove timeout to exit event loop

Let HTTP connections to timeout due to inactivity.
Let all remaning connections finish sending the response and close.

promag commented:
Split in multiple commits for easy review.

It can be arguable to remove the timeout but IMO it's safe because:
 - if an HTTP worker is with a long task then calling `event_base_loopbreak` is no good
 - active connections will timeout and close if no request/response happens.

This can be observed by:
```
bitcoind -regtest
time nc localhost 18443
bitcoin-cli -regtest stop
nc localhost 18443  0.00s user 0.01s system 0% cpu 30.025 total
```
The only way to improve this in a correct way is to lower the connection timeout.

ken2812221 reviewed (commented):
Concept ACK.

ken2812221 commented on src/httpserver.cpp:
```diff
@@ -579,6 +566,9 @@ void HTTPRequest::WriteHeader(const std::string& hdr, const std::string& value)
 void HTTPRequest::WriteReply(int nStatus, const std::string& strReply)
 {
     assert(!replySent && req);
+    if (ShutdownRequested()) {
```
To keep consistency and have lesser dependencies. Logically this should happen after http server interrupted.

promag replied:
Ah I see, I'll do it.

promag replied:
Actually I can't do that because that way the RPC `stop` reply won't get the `Connection: close` header — between `StartShutdown()` and `InterruptHTTPServer()` can go around 200ms:
https://github.com/bitcoin/bitcoin/blob/11e1ac3ae08535cefbd8235a8deb6cd100bcb2b1/src/bitcoind.cpp#L46-L54

ken2812221 replied:
Ok. Let's use ShutdownRequested for now. Then we can switch to the flag if #13211 merged.

promag replied:
I don't think that helps in this case.

ryanofsky commented:
Can you update the PR description to describe the change in external *behavior* instead of diving immediately into details of the code?

It would also help if each individual commit stated how it changes behavior, and why the change makes sense by itself or in conjunction with other changes.

promag commented:
> describe the change in external _behavior_

@ryanofsky sorry I don't understand what you mean with external behavior.

> and why the change makes sense by itself or in conjunction with other changes

I think I did?

> It would also help if each individual commit stated how it changes behavior

Will do.

promag commented:
@ryanofsky still testing but I believe it waits until HTTP request finishes or timeouts due to inactivity on the connection. In detail this means:
 - completes reading the request headers and body
 - worker computes full response and adds event to send the response
 - the response is sent and then the connection is finally closed
 - at this point, if it's the last connection, the event loop has no registered events and exists.

Updated commits and PR description. Added @ken2812221 suggestion.

promag committed:
rpc: Add wait argument to stop

promag committed:
qa: Test bitcond shutdown

promag commented:
- [x] 0ac433391 - qa: Expose get_rpc_proxy in TestNode
- [x] 331637d6c - rpc: Add wait argument to stop
- [ ] 9edfa116d - http: Send "Connection: close" header if shutdown is requested
- [ ] 8a937c1e5 - http: Unlisten sockets after all workers quit
- [ ] 0ae44271c - http: Remove unnecessary event_base_loopexit call
- [ ] bb6ea759e - http: Remove timeout to exit event loop
- [x] ab4f7a07c - qa: Test shutdown

@laanwj if you build only the marked commits then test `test/functional/feature_shutdown.py` fails.

ken2812221 commented:
The test can be fixed by github.com/ken2812221/bitcoin/tree/pr.14670

promag commented:
Thanks @ken2812221, I was going to try something along that after inspecting `stop_node` implementation.

@laanwj in order to properly test the fix, and to trigger the "race to exit the event loop", I had to add `wait` argument to `stop` RPC as explained in https://github.com/bitcoin/bitcoin/pull/13501#issuecomment-398219913.

promag commented on test/functional/feature_shutdown.py:
```diff
@@ -0,0 +1,30 @@
+
+    def run_test(self):
+        node = get_rpc_proxy(self.nodes[0].url, 1, timeout=600, coveragedir=self.nodes[0].coverage_dir)
+        t = Thread(target=waitfornewblock, args=(node,))
```
Thanks, I've changed a bit the code and now I don't think this applies, please check again.

laanwj commented:
@promag Makes sense, thanks for adding a test!

laanwj commented on src/rpc/server.cpp:
```diff
@@ -224,11 +224,20 @@ UniValue stop(const JSONRPCRequest& jsonRequest)
+            "stop ( wait )\n"
+            "\nStop Bitcoin server.\n"
+            "\nArguments:\n"
+            "1. \"wait\"     (int, optional, default=0) Milliseconds to wait before returning to the client\n"
```
maybe mention that this is for testing purposes

promag replied:
I also don't see reason to use this argument.

ken2812221 commented:
tACK 92a81adbe857b18537ad7d309c444dc818b1d2e8

promag commented:
Sorry @ken2812221, just pushed what @laanwj suggested. I'll squash later.

promag commented:
Rebased.

promag commented on src/rpc/server.cpp:
```diff
@@ -230,11 +230,23 @@ UniValue stop(const JSONRPCRequest& jsonRequest)
+                }}
+                .ToString() +
+            "\nArguments:\n"
+            "1. \"wait\"     (int, optional, default=0) Milliseconds to wait before returning to the client (intended for testing)\n"
```
I see, thanks.

promag replied:
Done.

ken2812221 reviewed (approved):
tACK 28479f926f21f2a91bec5a06671c60e5b0c55532

laanwj commented:
utACK 28479f926f21f2a91bec5a06671c60e5b0c55532

laanwj commented:
utACK 28479f926f21f2a91bec5a06671c60e5b0c55532

laanwj commented:
utACK 28479f926f21f2a91bec5a06671c60e5b0c55532

ken2812221 commented:
Is this another HTTP server bug?
https://travis-ci.org/bitcoin/bitcoin/jobs/467883032

```
stdout:
2018-12-14T08:34:22.452000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20181214_082355/feature_shutdown_0
2018-12-14T08:34:24.197000Z TestFramework (INFO): Stopping nodes
2018-12-14T08:34:24.197000Z TestFramework (INFO): Cleaning up /tmp/test_runner_₿_🏃_20181214_082355/feature_shutdown_0 on exit
2018-12-14T08:34:24.197000Z TestFramework (INFO): Tests successful
stderr:
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/feature_shutdown.py", line 12, in test_long_call
    block = node.waitfornewblock()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 134, in __call__
    response = self._request('POST', self.__url.path, postdata.encode('utf-8'))
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 105, in _request
    return self._get_response()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 166, in _get_response
    'code': -342, 'message': 'non-JSON HTTP response with \'%i %s\' from server' % (http_response.status, http_response.reason)})
test_framework.authproxy.JSONRPCException: non-JSON HTTP response with '503 Service Unavailable' from server (-342)
```

promag commented:
At least it doesn't look. Looks more a race between the thread that runs `test_long_call` and `stop_node` — if `stop_node` runs before then RPC would be unavailable. It seems unlikely because of the 1 second wait but it should possible considering the probable load.

promag commented:
Tested the change:
```diff
diff --git a/test/functional/feature_shutdown.py b/test/functional/feature_shutdown.py
index 5a0c88546..b0df81736 100755
--- a/test/functional/feature_shutdown.py
+++ b/test/functional/feature_shutdown.py
@@ -7,8 +7,11 @@
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import assert_equal, get_rpc_proxy
 from threading import Thread
+from time import sleep
+

 def test_long_call(node):
+    sleep(1)
     block = node.waitfornewblock()
     assert_equal(block['height'], 0)
```

which gives the following error:
```
ConnectionRefusedError: [Errno 61] Connection refused
```

Then on top of that:
```diff
@@ -20,6 +23,8 @@ class ShutdownTest(BitcoinTestFramework):

     def run_test(self):
         node = get_rpc_proxy(self.nodes[0].url, 1, timeout=600, coveragedir=self.nodes[0].coverage_dir)
+        # force connection establishment
+        node.getblockcount()
```

the error is now
```
test_framework.authproxy.JSONRPCException: non-JSON HTTP response with '503 Service Unavailable' from server (-342)
```
This could be easily fixed if we had a way to inspect how many RPC workers are running, for instance, by having a `getrpcinfo`.

Meanwhile this can be "prevented" by adding some relaxation:
```diff
@@ -26,6 +26,7 @@ class ShutdownTest(BitcoinTestFramework):
         # force connection establishment
         node.getblockcount()
         Thread(target=test_long_call, args=(node,)).start()
+        sleep(1)
```
